Marlin/Marlin_main.cpp

```diff
@@ -5428,55 +5428,73 @@ inline void gcode_M303() {
 
 #endif // SCARA
 
-#if ENABLED(EXT_SOLENOID)
-
-  void enable_solenoid(uint8_t num) {
-    switch(num) {
-      case 0:
-        OUT_WRITE(SOL0_PIN, HIGH);
-        break;
-        #if HAS_SOLENOID_1
-          case 1:
-            OUT_WRITE(SOL1_PIN, HIGH);
-            break;
-        #endif
-        #if HAS_SOLENOID_2
-          case 2:
-            OUT_WRITE(SOL2_PIN, HIGH);
-            break;
-        #endif
-        #if HAS_SOLENOID_3
-          case 3:
-            OUT_WRITE(SOL3_PIN, HIGH);
-            break;
-        #endif
-      default:
-        SERIAL_ECHO_START;
-        SERIAL_ECHOLNPGM(MSG_INVALID_SOLENOID);
-        break;
-    }
-  }
-
-  void enable_solenoid_on_active_extruder() { enable_solenoid(active_extruder); }
+#if ENABLED(PNEUMATICS)
 
   void disable_all_solenoids() {
-    OUT_WRITE(SOL0_PIN, LOW);
-    OUT_WRITE(SOL1_PIN, LOW);
-    OUT_WRITE(SOL2_PIN, LOW);
-    OUT_WRITE(SOL3_PIN, LOW);
+    #if HAS_SOLENOID_0
+      OUT_WRITE(SOL0_PIN, LOW);
+    #endif
+    #if HAS_SOLENOID_1
+      OUT_WRITE(SOL1_PIN, LOW);
+    #endif
   }
 
   /**
    * M380: Enable solenoid on the active extruder
    */
-  inline void gcode_M380() { enable_solenoid_on_active_extruder(); }
+  inline void gcode_M380() { 
+    int8_t current_solenoid_pin = -1;
+    switch(active_extruder) {
+      #if HAS_SOLENOID_0
+        case 0:
+          OUT_WRITE(SOL0_PIN, HIGH);
+          current_solenoid_pin = SOL0_PIN;
+          break;
+      #endif
+      #if HAS_SOLENOID_1
+        case 1:
+          OUT_WRITE(SOL1_PIN, HIGH);
+          current_solenoid_pin = SOL1_PIN;
+          break;
+      #endif
```

Comment:
Same question: Should SOLENOID_2 and SOLENOID_3 also be handled here?

If the answer is no I think their definitions should be removed from Conditionals.h above